compiler/rustc_ast/src/ast_like.rs

```diff
@@ -165,23 +192,32 @@ macro_rules! derive_has_attrs_no_tokens {
 macro_rules! derive_has_tokens_no_attrs {
     ($($ty:path),*) => { $(
         impl AstLike for $ty {
+            const SUPPORTS_CUSTOM_INNER_ATTRS: bool = false;
+
             fn attrs(&self) -> &[Attribute] {
                 &[]
             }
 
             fn visit_attrs(&mut self, _f: impl FnOnce(&mut Vec<Attribute>)) {}
-
             fn tokens_mut(&mut self) -> Option<&mut Option<LazyTokenStream>> {
                 Some(&mut self.tokens)
             }
         }
     )* }
 }
 
-// These AST nodes support both inert and active
-// attributes, so they also have tokens.
+// These ast nodes support both active and inert attributes,
+// so they have tokens collected to pass to proc macros
+derive_has_tokens_and_attrs! {
+    // Both `Item` and `AssocItem` can have bodies, which
+    // can contain inner attributes
+    const SUPPORTS_CUSTOM_INNER_ATTRS: bool = true;
+    Item, AssocItem, ForeignItem
+}
+
 derive_has_tokens_and_attrs! {
-    Item, Expr, Local, AssocItem, ForeignItem
+    const SUPPORTS_CUSTOM_INNER_ATTRS: bool = false;
+    Local, MacCallStmt, Expr
```

Comment:
Block expressions (`{}`) or other expressions with blocks like `for` or `async` still support inner macro attributes.
How are they treated?

All such expressions can be detected with 1-token lookahead if necessary.

Reply:
Oh - I actually banned inner attributes in *all* expressions (including `for`, `async`, and accidentally plain block expressions `{}`) in commit https://github.com/rust-lang/rust/commit/b71a50687d42ad559c6bcfbbf7952805bd6447e3. I guess I misunderstood your previous comment.

This complicates things, because in a case like `for _ in &[true] { #![macro_attr] }`, the inner attribute is actually part of the entire `for` expression. This means that we cannot bail out early when parsing a `for` expression, because we don't know if it will have inner attributes before we get to the block.

Reply:
Ok, we can probably ban all of them as a temporary measure, they are all feature gated anyway.

They can be enabled later in a separate PR - blocks, `for`s etc are relatively rare compared to expressions in general, so they shouldn't affect performance much, and they can be detected with 1-token lookahead.